packages/medusa/src/services/product.ts

```diff
@@ -0,0 +1,1067 @@
+      const optionRepo = manager.getCustomRepository(
+        this.productOptionRepository_
+      )
+
+      const { options, tags, type, images, ...rest } = productObject
+
+      if (!rest.thumbnail && images && images.length) {
+        rest.thumbnail = images[0]
+      }
+
+      // if product is a giftcard, we should disallow discounts
+      if (rest.is_giftcard) {
+        rest.discountable = false
+      }
+
+      try {
+        let product = productRepo.create(rest)
+
+        if (images) {
+          product.images = await imageRepo.upsertImages(images)
+        }
+
+        if (tags) {
+          product.tags = await productTagRepo.upsertTags(tags)
+        }
+
+        if (typeof type !== `undefined`) {
+          product.type_id = (await productTypeRepo.upsertType(type))?.id || null
+        }
+
+        product = await productRepo.save(product)
+
+        product.options = await Promise.all(
+          (options ?? []).map(async (option) => {
+            const res = optionRepo.create({ ...option, product_id: product.id })
+            await optionRepo.save(res)
+            return res
+          })
+        )
+
+        const result = await this.retrieve(product.id, {
+          relations: ["options"],
+        })
+
+        await this.eventBus_
+          .withTransaction(manager)
+          .emit(ProductService.Events.CREATED, {
+            id: result.id,
+          })
+        return result
+      } catch (error) {
+        throw formatException(error)
+      }
+    })
+  }
+
+  /**
+   * Updates a product. Product variant updates should use dedicated methods,
+   * e.g. `addVariant`, etc. The function will throw errors if metadata or
+   * product variant updates are attempted.
+   * @param {string} productId - the id of the product. Must be a string that
+   *   can be casted to an ObjectId
+   * @param {object} update - an object with the update values.
+   * @return {Promise} resolves to the update result.
+   */
+  async update(
+    productId: string,
+    update: UpdateProductInput
+  ): Promise<Product> {
+    return this.atomicPhase_(async (manager) => {
+      const productRepo = manager.getCustomRepository(this.productRepository_)
+      const productVariantRepo = manager.getCustomRepository(
+        this.productVariantRepository_
+      )
+      const productTagRepo = manager.getCustomRepository(
+        this.productTagRepository_
+      )
+      const productTypeRepo = manager.getCustomRepository(
+        this.productTypeRepository_
+      )
+      const imageRepo = manager.getCustomRepository(this.imageRepository_)
+
+      const product = await this.retrieve(productId, {
+        relations: ["variants", "tags", "images"],
+      })
+
+      const { variants, metadata, images, tags, type, ...rest } = update
+
+      if (!product.thumbnail && !update.thumbnail && images?.length) {
+        product.thumbnail = images[0]
+      }
+
+      if (images) {
```

Comment:
suggestion: Since the upsert method expect an array of value, if we have here an empty array the upsert method will do a query that get all the tuple that are in([]) which is not necessary since after that we will not loop because there is no element in the given array. Should we check `images?.length` as well here?